### PR TITLE
Support specialCaseMapEntries in incremental mode.

### LINKED ISSUE #697: specialCaseMapEntries does not work with incremental clutz
It appears we have special cased Map entries:
https://github.com/angular/clutz/blob/master/src/main/java/com/google/javascript/clutz/DeclarationGenerator.java#L2472

but that hack doesn't work any longer with incremental clutz. Needs revisiting, don't see obvious reason why it can't work with iclutz.

### PR DESCRIPTION
**Pull request by rkirov**
specialCaseMapEntries does some type inspection to make sure this is
indeed the special cased method. Those were write with the assumption
that externs are present, which is not true in incremental mode.

Same checks are now done using type names instead of resolved symbols.
There is no loss of safety.

Fixes #697

### CODE REVIEW

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 1]
```diff
@@ -2493,14 +2493,20 @@ private void emitProperty(
...
-      TemplatizedType ttType = propertyType.getReturnType().toMaybeTemplatizedType();
-      ImmutableList<JSType> ttypes = ttType.getTemplateTypes();
-      if (!isTemplateOf(ttType, "IteratorIterable")
+
+      ImmutableList<JSType> ttypes = null;
+      JSType returnType = propertyType.getReturnType();
+
+      // In incremental mode this type is noResolveType, while in total mode it is TemplatizedType
+      // They both have getTemplateTypes, but neither extends the other in the class heirarchy.
+      if (returnType.isTemplatizedType())
```
Comment (evmar) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: I think the rule is curlies if it spans more than one line.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 2]
```diff
@@ -2493,14 +2493,20 @@ private void emitProperty(
...
         return false;
       }
-      TemplatizedType ttType = propertyType.getReturnType().toMaybeTemplatizedType();
-      ImmutableList<JSType> ttypes = ttType.getTemplateTypes();
-      if (!isTemplateOf(ttType, "IteratorIterable")
+
+      ImmutableList<JSType> ttypes = null;
+      JSType returnType = propertyType.getReturnType();
+
+      // In incremental mode this type is noResolveType, while in total mode it is TemplatizedType
```
Comment (evmar) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: noResolvedType

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 3]
```diff
@@ -2493,14 +2493,20 @@ private void emitProperty(
...
+      // They both have getTemplateTypes, but neither extends the other in the class heirarchy.
+      if (returnType.isTemplatizedType())
+        ttypes = ((TemplatizedType) returnType).getTemplateTypes();
+      if (returnType.isNoResolvedType()) ttypes = ((NoResolvedType) returnType).getTemplateTypes();
```
Comment (evmar) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: Maybe "else if" here to make it clearer that it's checking a few props of returnType?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 1 above)
Comment (rkirov) [reply src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: of course. It used to be single line but googleJavaFormatter agressively puts everything on new line.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 2 above)
Comment (rkirov) [reply src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: done

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 3 above)
Comment (rkirov) [reply src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: done

**Pull request merged**